Implement auto-login and introduce SessionRepository for managing sessions

nielsvanvelzen opened the pull request:
**Changes**
This PR introduces the "SessionRepository" that manages the current session in the app. It contains two methods to restore the session based on the authentication preferences, these methods are called on application start.
When the session changes it uses a new class "ApiBinder" to configure the apiclient instance. I had to implement a hack to set the "currentUser" in the Application class (couldn't think of a better way unfortunatetely).

The StartupActivity now observes the current session, if a session is found it will wait for the currentUser to be assigned too and switch to the next activity. The splash screen is used to indicate something is happening. If the session is restored on start the startup activity will skip showing the user screen.

The sign out button on the home screen uses the destroyCurrentSession function which also unsets the currentuser.

**Issues**
Part of #612

thornbill reviewed (changes requested):
It kind of worked, but then I started getting this crash continuously.

<details>
<summary>Crash logs</summary>

```
2021-03-27 01:22:00.469 7450-7450/? I/androidtv.debu: Not late-enabling -Xcheck:jni (already on)
2021-03-27 01:22:00.500 7450-7450/org.jellyfin.androidtv.debug W/androidtv.debu: Unexpected CPU variant for X86 using defaults: x86
2021-03-27 01:22:00.940 7450-7450/org.jellyfin.androidtv.debug I/ACRA: ACRA is enabled for org.jellyfin.androidtv.debug, initializing...
2021-03-27 01:22:00.966 7450-7450/org.jellyfin.androidtv.debug W/androidtv.debu: Accessing hidden method Ldalvik/system/CloseGuard;->setEnabled(Z)V (light greylist, reflection)
2021-03-27 01:22:00.969 7450-7450/org.jellyfin.androidtv.debug I/JellyfinApplication: Application object created
2021-03-27 01:22:00.981 7450-7450/org.jellyfin.androidtv.debug I/[Koin]: [init] declare Android Context
2021-03-27 01:22:01.083 7450-7478/org.jellyfin.androidtv.debug E/WM-DelegatingWkrFctry: Unable to instantiate a ListenableWorker (org.jellyfin.androidtv.integration.LeanbackChannelWorker)
    org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.ListenableWorker' & qualifier:'org.jellyfin.androidtv.integration.LeanbackChannelWorker'. Check your definitions!
        at org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:282)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:251)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49)
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71)
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83)
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244)
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
2021-03-27 01:22:01.084 7450-7478/org.jellyfin.androidtv.debug E/AndroidRuntime: FATAL EXCEPTION: pool-2-thread-3
    Process: org.jellyfin.androidtv.debug, PID: 7450
    java.lang.Error: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.ListenableWorker' & qualifier:'org.jellyfin.androidtv.integration.LeanbackChannelWorker'. Check your definitions!
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1173)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.ListenableWorker' & qualifier:'org.jellyfin.androidtv.integration.LeanbackChannelWorker'. Check your definitions!
        at org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:282)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:251)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49)
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71)
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83)
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244)
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
2021-03-27 01:22:01.085 7450-7478/org.jellyfin.androidtv.debug E/ACRA: ACRA caught a Error for org.jellyfin.androidtv.debug
    java.lang.Error: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.ListenableWorker' & qualifier:'org.jellyfin.androidtv.integration.LeanbackChannelWorker'. Check your definitions!
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1173)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.ListenableWorker' & qualifier:'org.jellyfin.androidtv.integration.LeanbackChannelWorker'. Check your definitions!
        at org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:282)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:251)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49)
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71)
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83)
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244)
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
2021-03-27 01:22:01.096 7450-7450/org.jellyfin.androidtv.debug I/[Koin]: loaded 27 definitions - 3.82904 ms
2021-03-27 01:22:01.167 7450-7479/org.jellyfin.androidtv.debug E/[Koin]: Instance creation error : could not create instance for [Factory:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',qualifier:q:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',binds:androidx.work.ListenableWorker]: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.WorkerParameters'. Check your definitions!
    	org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:282)
    	org.koin.core.scope.Scope.resolveInstance(Scope.kt:251)
    	org.koin.core.scope.Scope.get(Scope.kt:204)
    	org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(AppModule.kt:68)
    	org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(Unknown Source:4)
    	org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:53)
    	org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:36)
    	org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:103)
    	org.koin.core.scope.Scope.resolveInstance(Scope.kt:236)
    	org.koin.core.scope.Scope.get(Scope.kt:204)
    	org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49)
    	androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71)
    	androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83)
    	androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244)
    	androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136)
    	androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
    	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
    	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
    	java.lang.Thread.run(Thread.java:764)
2021-03-27 01:22:01.169 7450-7479/org.jellyfin.androidtv.debug E/WM-DelegatingWkrFctry: Unable to instantiate a ListenableWorker (org.jellyfin.androidtv.integration.LeanbackChannelWorker)
    org.koin.core.error.InstanceCreationException: Could not create instance for [Factory:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',qualifier:q:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',binds:androidx.work.ListenableWorker]
        at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:62)
        at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:36)
        at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:103)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:236)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49)
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71)
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83)
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244)
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.WorkerParameters'. Check your definitions!
        at org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:282)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:251)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(AppModule.kt:68)
        at org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(Unknown Source:4)
        at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:53)
        at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:36) 
        at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:103) 
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:236) 
        at org.koin.core.scope.Scope.get(Scope.kt:204) 
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49) 
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71) 
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83) 
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244) 
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136) 
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
2021-03-27 01:22:01.173 7450-7479/org.jellyfin.androidtv.debug E/AndroidRuntime: FATAL EXCEPTION: pool-2-thread-4
    Process: org.jellyfin.androidtv.debug, PID: 7450
    java.lang.Error: org.koin.core.error.InstanceCreationException: Could not create instance for [Factory:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',qualifier:q:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',binds:androidx.work.ListenableWorker]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1173)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: org.koin.core.error.InstanceCreationException: Could not create instance for [Factory:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',qualifier:q:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',binds:androidx.work.ListenableWorker]
        at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:62)
        at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:36)
        at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:103)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:236)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49)
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71)
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83)
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244)
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
     Caused by: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.WorkerParameters'. Check your definitions!
        at org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:282)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:251)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(AppModule.kt:68)
        at org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(Unknown Source:4)
        at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:53)
        at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:36) 
        at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:103) 
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:236) 
        at org.koin.core.scope.Scope.get(Scope.kt:204) 
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49) 
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71) 
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83) 
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244) 
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136) 
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
2021-03-27 01:22:01.175 7450-7479/org.jellyfin.androidtv.debug E/ACRA: ACRA caught a Error for org.jellyfin.androidtv.debug
    java.lang.Error: org.koin.core.error.InstanceCreationException: Could not create instance for [Factory:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',qualifier:q:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',binds:androidx.work.ListenableWorker]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1173)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: org.koin.core.error.InstanceCreationException: Could not create instance for [Factory:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',qualifier:q:'org.jellyfin.androidtv.integration.LeanbackChannelWorker',binds:androidx.work.ListenableWorker]
        at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:62)
        at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:36)
        at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:103)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:236)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49)
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71)
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83)
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244)
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
     Caused by: org.koin.core.error.NoBeanDefFoundException: No definition found for class:'androidx.work.WorkerParameters'. Check your definitions!
        at org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:282)
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:251)
        at org.koin.core.scope.Scope.get(Scope.kt:204)
        at org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(AppModule.kt:68)
        at org.jellyfin.androidtv.di.AppModuleKt$appModule$1$7.invoke(Unknown Source:4)
        at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:53)
        at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:36) 
        at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:103) 
        at org.koin.core.scope.Scope.resolveInstance(Scope.kt:236) 
        at org.koin.core.scope.Scope.get(Scope.kt:204) 
        at org.koin.androidx.workmanager.factory.KoinWorkerFactory.createWorker(KoinWorkerFactory.kt:49) 
        at androidx.work.DelegatingWorkerFactory.createWorker(DelegatingWorkerFactory.java:71) 
        at androidx.work.WorkerFactory.createWorkerWithDefaultFallback(WorkerFactory.java:83) 
        at androidx.work.impl.WorkerWrapper.runWorker(WorkerWrapper.java:244) 
        at androidx.work.impl.WorkerWrapper.run(WorkerWrapper.java:136) 
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
2021-03-27 01:22:01.185 7450-7478/org.jellyfin.androidtv.debug I/Process: Sending signal. PID: 7450 SIG: 9
```

</details>

nielsvanvelzen commented:
That crash is not caused by this PR, I've had it happen before. Didn't look into it though.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/preference/AuthenticationPreferences.kt:
```diff
@@ -12,5 +12,10 @@ class AuthenticationPreferences(context: Context) : SharedPreferenceStore(
 		val systemUserId = Preference.string("system_user_id", "")
+
+		/**
+		 * Do not set direcltly use [SessionRepository] instead.
```
Fixed

nielsvanvelzen commented:
Found a relevant issue for the crash: https://github.com/InsertKoinIO/koin/issues/979
Should be fixed in Koin 3, although still in beta.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupActivity.kt:
```diff
@@ -65,15 +66,29 @@ class StartupActivity : FragmentActivity() {
+				application?.currentUserLiveData?.observe(this) { currentUser ->
+					Timber.i("CurrentUser changed to ${currentUser?.id} while waiting for startup.")
+
+					if (currentUser != null) openNextActivity()
```
I've noticed an issue with this part. Sometimes the `currentSession` is set but the app will never set the `currentUserLiveData` value (or it fails and sets it to null).

I'm not sure yet how to fix this but if I think of something I'll probably put it in a separate PR.

thornbill commented:
It looks like linting is failing due to some nullability issues, but looks good otherwise :+1:

nielsvanvelzen commented:
The linter issue was partially caused by me not using nullable types for the LiveData instances and partially a bug in the livedata library linter rules (but it was fixed in 2.3.1 (https://developer.android.com/jetpack/androidx/releases/lifecycle#version_231_2) which I updated to in #793). Fixed the types and rebased and linting passes now on my machine.
